### PR TITLE
Optimize preferred spreading for hostname topology

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Optimize spreading when using `kubernetes.io/hostname` as a topology key. We avoid calculating per-node counters during PreScore and do it in Score instead. This is faster because there are less map insertions and accesses.

```
default spreading

BenchmarkTestSelectorSpreadPriority/100nodes-56         	    5161	    226371 ns/op
BenchmarkTestSelectorSpreadPriority/1000nodes-56        	     788	   1516032 ns/op

master

BenchmarkTestDefaultEvenPodsSpreadPriority/100nodes-56         	    2168	    489570 ns/op
BenchmarkTestDefaultEvenPodsSpreadPriority/1000nodes-56        	     378	   3022404 ns/op

this PR

BenchmarkTestDefaultEvenPodsSpreadPriority/100nodes-56         	    3805	    329776 ns/op
BenchmarkTestDefaultEvenPodsSpreadPriority/1000nodes-56        	     631	   1858747 ns/op
```

This makes the implementation of default spreading with "topology spreading" closer to the default spreading:
- 0.3ms for 100 nodes, 1,000 pods (1.4x from default spreading)
- 1.8ms for 1,000 nodes, 10,000 pods (1.2x from default spreading)

**Which issue(s) this PR fixes**:

Part of #84936

**Special notes for your reviewer**:

The first commit contains the optimization.
The second commit updates the benchmark to reflect scheduler code, which parallelizes score calculation.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/priority important-longterm
/assign @Huang-Wei 

cc @ahg-g

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go [hunk 1]
```diff
@@ -82,3 +82,17 @@ func filterTopologySpreadConstraints(constraints []v1.TopologySpreadConstraint,
 	}
 	return result, nil
 }
+
+func countPodsMatchSelector(pods []*v1.Pod, selector labels.Selector, ns string) int {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/common.go]: nit: change int to int64 so we don't need to add explicit `int64(num)` conversion later.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go [hunk 2]
```diff
@@ -104,7 +107,7 @@ func (pl *PodTopologySpread) PreScore(
 	}
 
 	state := &preScoreState{
-		NodeNameSet:             sets.String{},
+		NodeNameSet:             make(sets.String, len(filteredNodes)),
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring.go]: Just for curiosity: how much percentage does it help in the performance boost?

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread_perf_test.go [hunk 3]
```diff
@@ -76,14 +77,13 @@ func BenchmarkTestSelectorSpreadPriority(b *testing.B) {
...
-				for _, node := range filteredNodes {
-					score, status := plugin.Score(ctx, state, pod, node.Name)
-					if !status.IsSuccess() {
-						b.Errorf("unexpected error: %v", status)
-					}
-					gotList = append(gotList, framework.NodeScore{Name: node.Name, Score: score})
+				gotList := make(framework.NodeScoreList, len(filteredNodes))
+				scoreNode := func(i int) {
+					n := filteredNodes[i]
+					score, _ := plugin.Score(context.Background(), state, pod, n.Name)
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread_perf_test.go]: Instantiating `context.Background()` everytime looks odd, cannot we reuse the outer `ctx`? Check out `wait.UntilWithContext`:

https://github.com/kubernetes/kubernetes/blob/fc8a39d439ea08b006fb7ea1db1058175cc7adf5/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L98

**Review by Huang-Wei — COMMENTED** [review]
Thanks @alculquicondor ! Just some nits, otherwise LGTM.

File: pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread_perf_test.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpodtopologyspread/default_pod_topology_spread_perf_test.go]: we can... I just missed it 😅

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/common.go]: I plan to reuse this in Filter, which uses `int32` for whatever reason. `int` seems more portable.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring.go]: The benchmark for 100 nodes remains at 0.3ms and goes up to 2.1ms for 1000 nodes.
The profile shows 10% CPU usage from `sets.String.Insert`, dominated by growing the hash table, which is not negligible.
I think it's fair to assume that if a user sets spreading, they expect all nodes to have the topology keys.

**Pull request merged**